Ensure the namespaces filtering is respected in packages.config package installation/updates in IVs installer API

erdembayar opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11002

Regression? Last working version: n/a

## Description
This change ensuring that the namespaces are respected during package installation in [IVs installer API](https://github.com/NuGet/Home/issues/10738#issuecomment-869887550). If there is same package Id exist in 2 repositories then it only request it from one it specified in namespace filter, previously we try to fetch from all possible repositories and install one from fastest which may not be what user really wanted.

This PR is continuation of https://github.com/NuGet/NuGet.Client/pull/4140.

I didn't manually test it exact way how Roslyn works, but I did add new unit tests for `InstallPackageApi`, `InstallLatestPackageApi` apis so I believe it should cover the scenario.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

erdembayar committed:
Add IVsPackageInstaller tests for Package namespace filter.

erdembayar committed:
Add Apex tests.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/IVsServicesTestCase.cs:
```diff
@@ -80,5 +82,237 @@ public void IVSPathContextProvider2_WithEmptySolution_WhenTryCreateUserWideConte
+            var privateRepositoryPath = Path.Combine(mainDirectory, "PrivateRepository");
+            Directory.CreateDirectory(privateRepositoryPath);
+
+            await CommonUtility.CreateNetCorePackageInSourceAsync(privateRepositoryPath, packageName, packageVersionV1, "Thisisfromprivaterepo1.txt");
```
The rest of the code looks reasonable, but I don't understand why we're creating netcore packages when all the projects can only be .NET Framework.

erdembayar replied:
Ok. Addressed it.

nkolev92 commented on test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs:
```diff
@@ -236,7 +236,7 @@ public async Task InstallPackageToProjectsFromUI_PackageNamespace_WithSingleFeed
             <namespace id=""Test.*"" />
         </packageSource>
     </packageNamespaces>
-//</configuration>");
```
were the `//`  causing test failures?

nkolev92 replied:
Or actually, why weren't they causing test failures.

erdembayar replied:
It's just typo from my previous PRs.

nkolev92 replied:
Doesn't that cause an unparseable config?

erdembayar replied:
It looks xml is forgiving for closing tags. It would definitely fail for opening tag error.

nkolev92 reviewed (commented):
It might be worth having a test where the source is specified.

erdembayar committed:
Address PR comment

erdembayar commented:
> It might be worth having a test where the source is specified.

@nkolev92 
Thank you for your review. I'll address sourceUri in different PR.
I addressed your comments, please review again.
